// @vscode-elements/elements/dist/vscode-table-header-cell/vscode-table-header-cell.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { html } from 'lit';
import { property } from 'lit/decorators.js';
import { customElement, VscElement } from '../includes/VscElement.js';
import styles from './vscode-table-header-cell.styles.js';
/**
 * @tag vscode-table-header-cell
 *
 * @cssprop [--vscode-foreground=#cccccc]
 * @cssprop [--vscode-font-family=sans-serif]
 * @cssprop [--vscode-font-size=13px]
 */
let VscodeTableHeaderCell = class VscodeTableHeaderCell extends VscElement {
    constructor() {
        super(...arguments);
        this.minWidth = '0';
        /** @internal */
        this.index = -1;
        /** @internal */
        this.role = 'columnheader';
    }
    willUpdate(changedProperties) {
        if (changedProperties.has('minWidth') && this.index > -1) {
            /** @internal */
            this.dispatchEvent(new CustomEvent('vsc-table-change-min-column-width', {
                detail: { columnIndex: this.index, propertyValue: this.minWidth },
                bubbles: true,
            }));
        }
    }
    render() {
        return html `
      <div class="wrapper">
        <slot></slot>
      </div>
    `;
    }
};
VscodeTableHeaderCell.styles = styles;
__decorate([
    property({ attribute: 'min-width' })
], VscodeTableHeaderCell.prototype, "minWidth", void 0);
__decorate([
    property({ type: Number })
], VscodeTableHeaderCell.prototype, "index", void 0);
__decorate([
    property({ reflect: true })
], VscodeTableHeaderCell.prototype, "role", void 0);
VscodeTableHeaderCell = __decorate([
    customElement('vscode-table-header-cell')
], VscodeTableHeaderCell);
export { VscodeTableHeaderCell };
