// @vscode-elements/elements/dist/vscode-tree-item/vscode-tree-item.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var VscodeTreeItem_1;
import { html, nothing } from 'lit';
import { consume } from '@lit/context';
import { property, queryAssignedElements, state } from 'lit/decorators.js';
import { classMap } from 'lit/directives/class-map.js';
import { customElement, VscElement } from '../includes/VscElement.js';
import { stylePropertyMap } from '../includes/style-property-map.js';
import { configContext, treeContext, } from '../vscode-tree/tree-context.js';
import { initPathTrackerProps } from '../vscode-tree/helpers.js';
import styles from './vscode-tree-item.styles.js';
import { ExpandMode, IndentGuides } from '../vscode-tree/vscode-tree.js';
const BASE_INDENT = 3;
const ARROW_CONTAINER_WIDTH = 30;
const arrowIcon = html `<svg
  width="16"
  height="16"
  viewBox="0 0 16 16"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    fill-rule="evenodd"
    clip-rule="evenodd"
    d="M10.072 8.024L5.715 3.667l.618-.62L11 7.716v.618L6.333 13l-.618-.619 4.357-4.357z"
  />
</svg>`;
function getParentItem(childItem) {
    if (!childItem.parentElement) {
        return null;
    }
    if (!(childItem.parentElement instanceof VscodeTreeItem)) {
        return null;
    }
    return childItem.parentElement;
}
/**
 * Represents an item in a Tree component.
 *
 * @tag vscode-tree
 *
 * @slot - Main content
 * @slot icon-branch - Custom icon for a closed branch item.
 * @slot icon-branch-opened - Custom icon for an opened branch item.
 * @slot icon-leaf - Custom icon for a leaf item.
 * @slot description - Description of the item. Displayed with a smaller font size and a less prominent color.
 * @slot actions - Container for action buttons.
 * @slot decoration - Container for small decorative elements aligned to the right edge of the item.
 */
let VscodeTreeItem = VscodeTreeItem_1 = class VscodeTreeItem extends VscElement {
    set selected(selected) {
        this._selected = selected;
        if (selected) {
            this._treeContextState.selectedItems.add(this);
        }
        else {
            this._treeContextState.selectedItems.delete(this);
        }
        this.ariaSelected = selected ? 'true' : 'false';
        this._updateActionsVisibility();
    }
    get selected() {
        return this._selected;
    }
    set path(newPath) {
        this._path = newPath;
    }
    get path() {
        return this._path;
    }
    //#endregion
    //#region lifecycle methods
    constructor() {
        super();
        //#region properties
        this.active = false;
        this.branch = false;
        this.hasActiveItem = false;
        this.hasSelectedItem = false;
        /** @internal */
        this.highlightedGuides = false;
        this.open = false;
        this.level = 0;
        this._selected = false;
        //#endregion
        //#region private variables
        this._path = [];
        this._hasBranchIcon = false;
        this._hasBranchOpenedIcon = false;
        this._hasLeafIcon = false;
        this._hasDescriptionSlotContent = false;
        this._hasActionsSlotContent = false;
        this._hasDecorationSlotContent = false;
        this._treeContextState = {
            isShiftPressed: false,
            selectedItems: new Set(),
            hoveredItem: null,
            allItems: null,
            itemListUpToDate: false,
            focusedItem: null,
            prevFocusedItem: null,
            hasBranchItem: false,
            rootElement: null,
            activeItem: null,
        };
        //#endregion
        //#region derived state
        this._isPointerInside = false;
        this._hasKeyboardFocus = false;
        this._handleMainSlotChange = () => {
            this._mainSlotChange();
            this._treeContextState.itemListUpToDate = false;
        };
        this._handleComponentFocus = () => {
            if (this._treeContextState.focusedItem &&
                this._treeContextState.focusedItem !== this) {
                if (!this._treeContextState.isShiftPressed) {
                    this._treeContextState.prevFocusedItem =
                        this._treeContextState.focusedItem;
                }
                this._treeContextState.focusedItem = null;
            }
            this._treeContextState.focusedItem = this;
        };
        this._handlePointerEnter = () => {
            this._isPointerInside = true;
            this._claimHover();
        };
        this._handlePointerLeave = (ev) => {
            this._isPointerInside = false;
            if (this._treeContextState.hoveredItem === this) {
                this._treeContextState.hoveredItem = null;
            }
            this._clearHoverState();
            const relatedTarget = ev.relatedTarget;
            if (relatedTarget instanceof Element) {
                const nextItem = relatedTarget.closest('vscode-tree-item');
                if (nextItem && nextItem !== this && nextItem.isConnected) {
                    nextItem._adoptHoverFromSibling();
                }
            }
        };
        this._handleFocusIn = () => {
            this._updateFocusState();
        };
        this._handleFocusOut = () => {
            this._updateFocusState();
        };
        this._internals = this.attachInternals();
        this.addEventListener('focus', this._handleComponentFocus);
        this.addEventListener('pointerenter', this._handlePointerEnter);
        this.addEventListener('pointerleave', this._handlePointerLeave);
        this.addEventListener('focusin', this._handleFocusIn);
        this.addEventListener('focusout', this._handleFocusOut);
    }
    connectedCallback() {
        super.connectedCallback();
        this._mainSlotChange();
        this.role = 'treeitem';
        this.ariaDisabled = 'false';
    }
    firstUpdated(changedProperties) {
        super.firstUpdated(changedProperties);
        this._refreshDescriptionSlotState();
        this._refreshActionsSlotState();
        this._refreshDecorationSlotState();
        if (this.matches(':hover')) {
            this._isPointerInside = true;
            this._claimHover();
        }
        else {
            this._updateActionsVisibility();
        }
    }
    willUpdate(changedProperties) {
        if (changedProperties.has('active')) {
            this._toggleActiveState();
        }
        if (changedProperties.has('open') || changedProperties.has('branch')) {
            this._setAriaExpanded();
        }
    }
    //#endregion
    //#region private methods
    _setAriaExpanded() {
        if (!this.branch) {
            this.ariaExpanded = null;
        }
        else {
            this.ariaExpanded = this.open ? 'true' : 'false';
        }
    }
    _setHasActiveItemFlagOnParent(childItem, value) {
        const parent = getParentItem(childItem);
        if (parent) {
            parent.hasActiveItem = value;
        }
    }
    _refreshDescriptionSlotState() {
        const hasContent = (this._descriptionSlotElements?.length ?? 0) > 0;
        this._hasDescriptionSlotContent = hasContent;
        this._setCustomState('has-description', hasContent);
    }
    _refreshActionsSlotState() {
        const hasContent = (this._actionsSlotElements?.length ?? 0) > 0;
        this._hasActionsSlotContent = hasContent;
        this._setCustomState('has-actions', hasContent);
        this._updateActionsVisibility();
    }
    _refreshDecorationSlotState() {
        const hasContent = (this._decorationSlotElements?.length ?? 0) > 0;
        const prevHasDecoration = this._hasDecorationSlotContent;
        this._hasDecorationSlotContent = hasContent;
        this._setCustomState('has-decoration', hasContent);
        if (prevHasDecoration !== hasContent) {
            this.requestUpdate();
        }
    }
    _setCustomState(stateName, present) {
        if (!this._internals?.states) {
            return;
        }
        try {
            if (present) {
                this._internals.states.add(stateName);
            }
            else {
                this._internals.states.delete(stateName);
            }
        }
        catch {
            // https://developer.mozilla.org/en-US/docs/Web/API/CustomStateSet#compatibility_with_dashed-ident_syntax
            if (present) {
                this._internals.states.add(`--${stateName}`);
            }
            else {
                this._internals.states.delete(`--${stateName}`);
            }
        }
    }
    _getActiveElement() {
        const root = this.getRootNode({ composed: true });
        if (root instanceof Document) {
            return root.activeElement instanceof Element ? root.activeElement : null;
        }
        if (root instanceof ShadowRoot) {
            return root.activeElement instanceof Element ? root.activeElement : null;
        }
        return null;
    }
    _isActiveElementInActions(activeElement) {
        if (!activeElement) {
            return false;
        }
        return (this._actionsSlotElements ?? []).some((element) => element === activeElement || element.contains(activeElement));
    }
    _updateActionsVisibility() {
        if (!this._hasActionsSlotContent) {
            this._setCustomState('show-actions', false);
            return;
        }
        const activeElement = this._getActiveElement();
        const isActionsFocused = this._isActiveElementInActions(activeElement);
        const shouldShow = this.selected ||
            this._isPointerInside ||
            this._hasKeyboardFocus ||
            isActionsFocused;
        this._setCustomState('show-actions', shouldShow);
    }
    _updateFocusState() {
        const hostFocusVisible = this.matches(':focus-visible');
        this._setCustomState('focus-visible', hostFocusVisible);
        const activeElement = this._getActiveElement();
        let owner = null;
        if (activeElement instanceof Element) {
            owner = activeElement.closest('vscode-tree-item');
            if (!owner) {
                const root = activeElement.getRootNode();
                if (root instanceof ShadowRoot && root.host instanceof VscodeTreeItem_1) {
                    owner = root.host;
                }
            }
        }
        const hasKeyboardFocus = owner === this;
        this._hasKeyboardFocus = hasKeyboardFocus;
        this._setCustomState('keyboard-focus', hasKeyboardFocus);
        this._updateActionsVisibility();
    }
    _clearHoverState() {
        this._isPointerInside = false;
        this._setCustomState('hover', false);
        this._updateActionsVisibility();
    }
    _adoptHoverFromSibling() {
        this._isPointerInside = true;
        this._claimHover();
    }
    _claimHover() {
        const treeState = this._treeContextState;
        if (treeState.hoveredItem && treeState.hoveredItem !== this) {
            treeState.hoveredItem._clearHoverState();
        }
        treeState.hoveredItem = this;
        this._setCustomState('hover', true);
        this._updateActionsVisibility();
    }
    _toggleActiveState() {
        if (this.active) {
            if (this._treeContextState.activeItem) {
                this._treeContextState.activeItem.active = false;
                this._setHasActiveItemFlagOnParent(this._treeContextState.activeItem, false);
            }
            this._treeContextState.activeItem = this;
            this._setHasActiveItemFlagOnParent(this, true);
            this.tabIndex = 0;
            this._setCustomState('active', true);
        }
        else {
            if (this._treeContextState.activeItem === this) {
                this._treeContextState.activeItem = null;
                this._setHasActiveItemFlagOnParent(this, false);
            }
            this.tabIndex = -1;
            this._setCustomState('active', false);
        }
    }
    _selectItem(isCtrlDown) {
        const { selectedItems } = this._treeContextState;
        const { multiSelect } = this._configContext;
        const prevSelected = new Set(selectedItems);
        if (multiSelect && isCtrlDown) {
            this.selected = !this.selected;
        }
        else {
            Array.from(selectedItems).forEach((li) => {
                if (li !== this) {
                    li.selected = false;
                }
            });
            selectedItems.clear();
            this.selected = true;
        }
        const affected = new Set([
            ...prevSelected,
            ...selectedItems,
        ]);
        affected.add(this);
        affected.forEach((li) => li._updateActionsVisibility());
    }
    _selectRange() {
        const prevFocused = this._treeContextState.prevFocusedItem;
        if (!prevFocused || prevFocused === this) {
            return;
        }
        const prevSelected = new Set(this._treeContextState.selectedItems);
        if (!this._treeContextState.itemListUpToDate) {
            this._treeContextState.allItems =
                this._treeContextState.rootElement.querySelectorAll('vscode-tree-item');
            if (this._treeContextState.allItems) {
                this._treeContextState.allItems.forEach((li, i) => {
                    li.dataset.score = i.toString();
                });
            }
            this._treeContextState.itemListUpToDate = true;
        }
        let from = +(prevFocused.dataset.score ?? -1);
        let to = +(this.dataset.score ?? -1);
        if (from > to) {
            [from, to] = [to, from];
        }
        Array.from(this._treeContextState.selectedItems).forEach((li) => (li.selected = false));
        this._treeContextState.selectedItems.clear();
        this._selectItemsAndAllVisibleDescendants(from, to);
        const affected = new Set([
            ...prevSelected,
            ...this._treeContextState.selectedItems,
        ]);
        affected.add(this);
        affected.forEach((li) => li._updateActionsVisibility());
    }
    _selectItemsAndAllVisibleDescendants(from, to) {
        let i = from;
        while (i <= to) {
            if (this._treeContextState.allItems) {
                const item = this._treeContextState.allItems[i];
                if (item.branch && !item.open) {
                    item.selected = true;
                    const numChildren = item.querySelectorAll('vscode-tree-item').length;
                    i += numChildren;
                }
                else if (item.branch && item.open) {
                    item.selected = true;
                    i += this._selectItemsAndAllVisibleDescendants(i + 1, to);
                }
                else {
                    item.selected = true;
                    i += 1;
                }
            }
        }
        return i;
    }
    _mainSlotChange() {
        this._initiallyAssignedTreeItems.forEach((li) => {
            li.setAttribute('slot', 'children');
        });
    }
    //#endregion
    //#region event handlers
    _handleChildrenSlotChange() {
        initPathTrackerProps(this, this._childrenTreeItems);
        if (this._treeContextState.rootElement) {
            this._treeContextState.rootElement.updateHasBranchItemFlag();
        }
    }
    _handleDescriptionSlotChange() {
        this._refreshDescriptionSlotState();
    }
    _handleActionsSlotChange() {
        this._refreshActionsSlotState();
    }
    _handleDecorationSlotChange() {
        this._refreshDecorationSlotState();
    }
    _handleContentClick(ev) {
        ev.stopPropagation();
        const isCtrlDown = ev.ctrlKey || ev.metaKey;
        const isShiftDown = ev.shiftKey;
        if (isShiftDown && this._configContext.multiSelect) {
            this._selectRange();
            this._treeContextState.emitSelectEvent?.();
            this.updateComplete.then(() => {
                this._treeContextState.highlightIndentGuides?.();
            });
        }
        else {
            this._selectItem(isCtrlDown);
            this._treeContextState.emitSelectEvent?.();
            this.updateComplete.then(() => {
                this._treeContextState.highlightIndentGuides?.();
            });
            if (this._configContext.expandMode === ExpandMode.singleClick) {
                if (this.branch && !(this._configContext.multiSelect && isCtrlDown)) {
                    this.open = !this.open;
                }
            }
        }
        this.active = true;
        if (!isShiftDown) {
            this._treeContextState.prevFocusedItem = this;
        }
    }
    _handleDoubleClick(ev) {
        if (this._configContext.expandMode === ExpandMode.doubleClick) {
            if (this.branch &&
                !(this._configContext.multiSelect && (ev.ctrlKey || ev.metaKey))) {
                this.open = !this.open;
            }
        }
    }
    _handleIconSlotChange(ev) {
        const slot = ev.target;
        const hasContent = slot.assignedElements().length > 0;
        switch (slot.name) {
            case 'icon-branch':
                this._hasBranchIcon = hasContent;
                break;
            case 'icon-branch-opened':
                this._hasBranchOpenedIcon = hasContent;
                break;
            case 'icon-leaf':
                this._hasLeafIcon = hasContent;
                break;
            default:
        }
    }
    //#endregion
    render() {
        const { hideArrows, indent, indentGuides } = this._configContext;
        const { hasBranchItem } = this._treeContextState;
        let indentation = BASE_INDENT + this.level * indent;
        const guideOffset = !hideArrows ? 13 : 3;
        const indentGuideX = BASE_INDENT + this.level * indent + guideOffset;
        if (!this.branch && !hideArrows && hasBranchItem) {
            indentation += ARROW_CONTAINER_WIDTH;
        }
        const hasVisibleIcon = (this._hasBranchIcon && this.branch) ||
            (this._hasBranchOpenedIcon && this.branch && this.open) ||
            (this._hasLeafIcon && !this.branch);
        const wrapperClasses = {
            wrapper: true,
            active: this.active,
            'has-description': this._hasDescriptionSlotContent,
            'has-actions': this._hasActionsSlotContent,
            'has-decoration': this._hasDecorationSlotContent,
        };
        const childrenClasses = {
            children: true,
            guide: indentGuides !== IndentGuides.none,
            'default-guide': indentGuides !== IndentGuides.none,
            'highlighted-guide': this.highlightedGuides,
        };
        const iconContainerClasses = {
            'icon-container': true,
            'has-icon': hasVisibleIcon,
        };
        const contentClasses = {
            content: true,
            'has-description': this._hasDescriptionSlotContent,
            'has-decoration': this._hasDecorationSlotContent,
        };
        return html ` <div class="root">
      <div
        class=${classMap(wrapperClasses)}
        part="wrapper"
        @click=${this._handleContentClick}
        @dblclick=${this._handleDoubleClick}
        .style=${stylePropertyMap({ paddingLeft: `${indentation}px` })}
      >
        ${this.branch && !hideArrows
            ? html `<div
              class=${classMap({
                'arrow-container': true,
                'icon-rotated': this.open,
            })}
              part="arrow-icon-container"
            >
              ${arrowIcon}
            </div>`
            : nothing}
        <div class=${classMap(iconContainerClasses)} part="icon-container">
          ${this.branch && !this.open
            ? html `<slot
                name="icon-branch"
                @slotchange=${this._handleIconSlotChange}
              ></slot>`
            : nothing}
          ${this.branch && this.open
            ? html `<slot
                name="icon-branch-opened"
                @slotchange=${this._handleIconSlotChange}
              ></slot>`
            : nothing}
          ${!this.branch
            ? html `<slot
                name="icon-leaf"
                @slotchange=${this._handleIconSlotChange}
              ></slot>`
            : nothing}
        </div>
        <div class=${classMap(contentClasses)} part="content">
          <span class="label" part="label">
            <slot @slotchange=${this._handleMainSlotChange}></slot>
          </span>
          <span
            class="description"
            part="description"
            ?hidden=${!this._hasDescriptionSlotContent}
          >
            <slot
              name="description"
              @slotchange=${this._handleDescriptionSlotChange}
            ></slot>
          </span>
          <div class="actions" part="actions">
            <slot
              name="actions"
              @slotchange=${this._handleActionsSlotChange}
            ></slot>
          </div>
          <div class="decoration" part="decoration">
            <slot
              name="decoration"
              @slotchange=${this._handleDecorationSlotChange}
            ></slot>
          </div>
        </div>
      </div>
      <div
        class=${classMap(childrenClasses)}
        .style=${stylePropertyMap({
            '--indentation-guide-left': `${indentGuideX}px`,
        })}
        role="group"
        part="children"
      >
        <slot
          name="children"
          @slotchange=${this._handleChildrenSlotChange}
        ></slot>
      </div>
    </div>`;
    }
};
VscodeTreeItem.styles = styles;
__decorate([
    property({ type: Boolean })
], VscodeTreeItem.prototype, "active", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], VscodeTreeItem.prototype, "branch", void 0);
__decorate([
    property({ type: Boolean })
], VscodeTreeItem.prototype, "hasActiveItem", void 0);
__decorate([
    property({ type: Boolean })
], VscodeTreeItem.prototype, "hasSelectedItem", void 0);
__decorate([
    property({ type: Boolean })
], VscodeTreeItem.prototype, "highlightedGuides", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], VscodeTreeItem.prototype, "open", void 0);
__decorate([
    property({ type: Number, reflect: true })
], VscodeTreeItem.prototype, "level", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], VscodeTreeItem.prototype, "selected", null);
__decorate([
    state()
], VscodeTreeItem.prototype, "_hasBranchIcon", void 0);
__decorate([
    state()
], VscodeTreeItem.prototype, "_hasBranchOpenedIcon", void 0);
__decorate([
    state()
], VscodeTreeItem.prototype, "_hasLeafIcon", void 0);
__decorate([
    state()
], VscodeTreeItem.prototype, "_hasDescriptionSlotContent", void 0);
__decorate([
    state()
], VscodeTreeItem.prototype, "_hasActionsSlotContent", void 0);
__decorate([
    state()
], VscodeTreeItem.prototype, "_hasDecorationSlotContent", void 0);
__decorate([
    consume({ context: treeContext, subscribe: true })
], VscodeTreeItem.prototype, "_treeContextState", void 0);
__decorate([
    consume({ context: configContext, subscribe: true })
], VscodeTreeItem.prototype, "_configContext", void 0);
__decorate([
    queryAssignedElements({ selector: 'vscode-tree-item' })
], VscodeTreeItem.prototype, "_initiallyAssignedTreeItems", void 0);
__decorate([
    queryAssignedElements({ selector: 'vscode-tree-item', slot: 'children' })
], VscodeTreeItem.prototype, "_childrenTreeItems", void 0);
__decorate([
    queryAssignedElements({ slot: 'description', flatten: true })
], VscodeTreeItem.prototype, "_descriptionSlotElements", void 0);
__decorate([
    queryAssignedElements({ slot: 'actions', flatten: true })
], VscodeTreeItem.prototype, "_actionsSlotElements", void 0);
__decorate([
    queryAssignedElements({ slot: 'decoration', flatten: true })
], VscodeTreeItem.prototype, "_decorationSlotElements", void 0);
VscodeTreeItem = VscodeTreeItem_1 = __decorate([
    customElement('vscode-tree-item')
], VscodeTreeItem);
export { VscodeTreeItem };
