// @vscode-elements/elements/dist/vscode-table/ColumnResizeController.js
import { percent, px, toPercent, toPx } from '../includes/sizes.js';
import { calculateColumnWidths } from './calculations.js';
export class ColumnResizeController {
    constructor(host) {
        this._hostWidth = px(0);
        this._hostX = px(0);
        this._activeSplitter = null;
        this._columnMinWidths = new Map();
        this._columnWidths = [];
        this._dragState = null;
        this._cachedSplitterPositions = null;
        (this._host = host).addController(this);
    }
    hostConnected() {
        this.saveHostDimensions();
    }
    get isDragging() {
        return this._dragState !== null;
    }
    get splitterPositions() {
        if (this._cachedSplitterPositions) {
            return this._cachedSplitterPositions;
        }
        const result = [];
        let acc = percent(0);
        for (let i = 0; i < this._columnWidths.length - 1; i++) {
            acc = percent(acc + this._columnWidths[i]);
            result.push(acc);
        }
        this._cachedSplitterPositions = result;
        return result;
    }
    getActiveSplitterCalculatedPosition() {
        const splitterPositions = this.splitterPositions;
        if (!this._dragState) {
            return px(0);
        }
        const activeSplitterPos = splitterPositions[this._dragState.splitterIndex];
        const activeSplitterPosPx = this._toPx(activeSplitterPos);
        return activeSplitterPosPx;
    }
    get columnWidths() {
        return this._columnWidths;
    }
    get columnMinWidths() {
        return new Map(this._columnMinWidths);
    }
    saveHostDimensions() {
        const cr = this._host.getBoundingClientRect();
        const { width, x } = cr;
        this._hostWidth = px(width);
        this._hostX = px(x);
        return this;
    }
    setActiveSplitter(splitter) {
        this._activeSplitter = splitter;
        return this;
    }
    getActiveSplitter() {
        return this._activeSplitter;
    }
    setColumnMinWidthAt(colIndex, value) {
        this._columnMinWidths.set(colIndex, value);
        this._host.requestUpdate();
        return this;
    }
    setColumWidths(widths) {
        this._columnWidths = widths;
        this._cachedSplitterPositions = null;
        this._host.requestUpdate();
        return this;
    }
    shouldDrag(event) {
        return (+event.currentTarget.dataset.index ===
            this._dragState?.splitterIndex);
    }
    startDrag(event) {
        event.stopPropagation();
        if (this._dragState) {
            return;
        }
        this._activeSplitter?.setPointerCapture(event.pointerId);
        const mouseX = event.pageX;
        const splitter = event.currentTarget;
        const splitterX = splitter.getBoundingClientRect().x;
        const xOffset = px(mouseX - splitterX);
        this._dragState = {
            dragOffset: px(xOffset),
            pointerId: event.pointerId,
            splitterIndex: +splitter.dataset.index,
            prevX: px(mouseX - xOffset),
        };
        this._host.requestUpdate();
    }
    drag(event) {
        event.stopPropagation();
        if (!event?.currentTarget?.hasPointerCapture?.(event.pointerId)) {
            return;
        }
        if (!this._dragState) {
            return;
        }
        if (event.pointerId !== this._dragState.pointerId) {
            return;
        }
        if (!this.shouldDrag(event)) {
            return;
        }
        const mouseX = event.pageX;
        const x = px(mouseX - this._dragState.dragOffset);
        const deltaPx = px(x - this._dragState.prevX);
        const delta = this._toPercent(deltaPx);
        this._dragState.prevX = x;
        const splitterPos = this.getActiveSplitterCalculatedPosition();
        if ((deltaPx <= 0 && mouseX > splitterPos + this._hostX) ||
            (deltaPx > 0 && mouseX < splitterPos + this._hostX)) {
            return;
        }
        this._columnWidths = calculateColumnWidths(this._columnWidths, this._dragState.splitterIndex, delta, this._columnMinWidths);
        this._cachedSplitterPositions = null;
        this._host.requestUpdate();
    }
    stopDrag(event) {
        event.stopPropagation();
        if (!this._dragState) {
            return;
        }
        const el = event.currentTarget;
        try {
            el.releasePointerCapture(this._dragState.pointerId);
        }
        catch (e) {
            // ignore
        }
        this._dragState = null;
        this._activeSplitter = null;
        this._host.requestUpdate();
    }
    _toPercent(px) {
        return toPercent(px, this._hostWidth);
    }
    _toPx(percent) {
        return toPx(percent, this._hostWidth);
    }
}
